parley/src/layout/alignment.rs

```diff
@@ -32,37 +34,43 @@ pub(crate) fn align<B: Brush>(
             continue;
         }
 
-        match alignment {
-            Alignment::Start => {
+        match (alignment, is_rtl) {
+            (Alignment::Left, _) | (Alignment::Start, false) | (Alignment::End, true) => {
                 // Do nothing
             }
-            Alignment::End => {
+            (Alignment::Right, _) | (Alignment::Start, true) | (Alignment::End, false) => {
                 line.metrics.offset = free_space;
             }
-            Alignment::Middle => {
+            (Alignment::Middle, _) => {
                 line.metrics.offset = free_space * 0.5;
             }
-            Alignment::Justified => {
+            (Alignment::Justified, _) => {
                 // Justified alignment doesn't have any effect if free_space is negative or zero
                 if free_space <= 0.0 {
                     continue;
                 }
 
-                // Justified alignment doesn't apply to the last line of a paragraph (`BreakReason::None`)
-                // or if there are no whitespace gaps to adjust
+                // Justified alignment doesn't apply to the last line of a paragraph
+                // (`BreakReason::None`) or if there are no whitespace gaps to adjust. In that
+                // case, start-align, i.e., left-align for LTR text and right-align for RTL text.
                 if line.break_reason == BreakReason::None || line.num_spaces == 0 {
+                    if is_rtl {
+                        line.metrics.offset = free_space;
+                    }
                     continue;
                 }
 
                 let adjustment = free_space / line.num_spaces as f32;
                 let mut applied = 0;
-                for line_item in layout.line_items[line.item_range.clone()]
-                    .iter()
-                    .filter(|item| item.is_text_run())
-                {
-                    // Iterate over clusters in the run
-                    //   - Iterate forwards for even bidi levels (which represent LTR runs)
-                    //   - Iterate backwards for odd bidi levels (which represent RTL runs)
+                // Iterate over text runs in the line and clusters in the text run
+                //   - Iterate forwards for even bidi levels (which represent LTR runs)
+                //   - Iterate backwards for odd bidi levels (which represent RTL runs)
+                let line_items: &mut dyn Iterator<Item = &LineItemData> = if is_rtl {
+                    &mut layout.line_items[line.item_range.clone()].iter().rev()
+                } else {
+                    &mut layout.line_items[line.item_range.clone()].iter()
+                };
+                for line_item in line_items.filter(|item| item.is_text_run()) {
```

Comment:
I like this trick with a trait object, and it might be applicable in other places that deal with bidi as well. However, and this is probably _very_ premature optimization, but the Rust compiler doesn't optimize this very well (https://godbolt.org/z/sMqsTdfMs). Obviously, without benchmarks it is impossible to say what the actual impact is, but I thought I'd mention it at least.

Reply:
You are right about that, Rust calls through the vtable and unfortunately doesn't know to inline it with a branch. I think ideally this wouldn't be here. It may be possible with e.g. `skip`, but it's probably a good idea to first have tests with mixed LTR/RTL text working.

Reply:
I believe that you can use `Iterator::for_each`, which should optimise much more reasonably.

Reply:
You could do the same for the inner loops over the clusters, because they are essentially duplicated at the moment.

Reply:
I believe that (in theory) `Iterator::fold` should optimise even better, at the cost of not being able to `break` when `num_spaces` is reached (but I'm guessing that `break` isn't saving us much on typical text).

Reply:
I've opened https://github.com/linebender/parley/issues/261 about creating benchmarks so that we can actually measure the impact of these kind of changes.